# Fuel Management Project

## Overview

This project aims to analyze and optimize fuel usage within a specified system. The primary objectives include data cleaning, preprocessing, exploratory data analysis (EDA), and the development of predictive models to improve fuel management efficiency.

## Contents

1. **Data Cleaning and Preprocessing**: Steps to handle missing values, remove duplicates, and normalize data.
3. **Feature Engineering**: Creation of new features or transformation of existing ones to improve model performance.
5. **Model Evaluation**: Assessment of model performance using appropriate metrics and validation techniques.
6. **Conclusions and Insights**: Summary of findings and recommendations for improving fuel management.


# 1. Data Loading and Overview

In [1]:
#1.1: Importing the essential Libraries
import pandas as pd
import numpy as np
import math
import matplotlib.pyplot as plt
import seaborn as sns

In [4]:
#1.2: Loading the data
df = pd.read_csv("Final Data.csv")
df.head(10)

Number Registration                              Details Product Odometer  \
0    NaN          NaN                                  NaN     NaN      NaN   
1   0001           U1  Jaguar XJL Portfolio V6 D Limousine  Diesel    67691   
2   0012      FL65FBA             Vauxhall CORSA  CDTI Car  Diesel     9329   
3   0013      MJ16XMY   Vauxhall Mokka 4x4 Utility Vehicle  Diesel    29064   
4   0016      BP16MKK                      Nissan Note Car  Diesel    42233   
5   0017      YD16NAL                      Peugeot car Car  Diesel    20475   
6   0018      YD16NAK                  Peugeot 308 Car Car  Diesel   115020   
7   0019      BP16OPD                      Nissan Note Car  Diesel    15559   
8   0020      YH17VCF      Peugeot Expert Combi 9 seat MPV  Diesel    37352   
9   0021      MF68FNR             Renault  Traffic Welfare  Diesel      112   

  Distance   Unit     MPG  
0      NaN    NaN     NaN  
1        0  Miles    0.00  
2     2849  Miles   50.04  
3     5369  Miles   39.29  
4     1052  Miles   56.17  
5     3319  Miles  108.32  
6    36100  Miles   50.79  
7      948  Miles   36.04  
8    14202  Miles   34.75  
9        0  Miles    0.00

In [5]:
# 1.3: Looking for data types
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 1490 entries, 0 to 1489
Data columns (total 8 columns):
 #   Column        Non-Null Count  Dtype  
---  ------        --------------  -----  
 0   Number        1486 non-null   object 
 1   Registration  1483 non-null   object 
 2   Details       1484 non-null   object 
 3   Product       1484 non-null   object 
 4   Odometer      1484 non-null   object 
 5   Distance      1484 non-null   object 
 6   Unit          1482 non-null   object 
 7   MPG           1482 non-null   float64
dtypes: float64(1), object(7)
memory usage: 93.3+ KB


## Dataset Overview

The dataset contains 1,490 entries with 8 columns related to fuel management. Below is a brief description of the dataset structure:

- **Number**: Identification number for each entry (1486 non-null values, dtype: object)
- **Registration**: Vehicle registration details (1483 non-null values, dtype: object)
- **Details**: Additional details about the entry (1484 non-null values, dtype: object)
- **Product**: Type of fuel product used (1484 non-null values, dtype: object)
- **Odometer**: Odometer reading at the time of fuel entry (1484 non-null values, dtype: object)
- **Distance**: Distance covered (1484 non-null values, dtype: object)
- **Unit**: Unit of measurement for fuel (1482 non-null values, dtype: object)
- **MPG**: Miles per gallon (1482 non-null values, dtype: float64)

The dataset has some missing values across several columns, which will be addressed during the data cleaning and preprocessing phase.

In [7]:
#1.4: Statistical significance of the data
df.describe(include = 'all')

Number Registration                    Details Product Odometer  \
count    1486         1483                       1484    1484     1484   
unique   1485         1452                        301       5     1359   
top      1482      MW16DDL  Peugeot Partner Van Small  Diesel        0   
freq        2            2                        104    1440      114   
mean      NaN          NaN                        NaN     NaN      NaN   
std       NaN          NaN                        NaN     NaN      NaN   
min       NaN          NaN                        NaN     NaN      NaN   
25%       NaN          NaN                        NaN     NaN      NaN   
50%       NaN          NaN                        NaN     NaN      NaN   
75%       NaN          NaN                        NaN     NaN      NaN   
max       NaN          NaN                        NaN     NaN      NaN   

       Distance   Unit          MPG  
count      1484   1482  1482.000000  
unique     1187      4          NaN  
top           0  Miles          NaN  
freq        250   1265          NaN  
mean        NaN    NaN    12.037726  
std         NaN    NaN   246.914654  
min         NaN    NaN -9238.620000  
25%         NaN    NaN     3.167500  
50%         NaN    NaN    19.750000  
75%         NaN    NaN    29.885000  
max         NaN    NaN   824.850000

### Summary Statistics

- **Number**
  - Count: 1486
  - Unique: 1485
  - Most frequent (top): 1482 (2 occurrences)

- **Registration**
  - Count: 1483
  - Unique: 1452
  - Most frequent (top): MW16DDL (2 occurrences)

- **Details**
  - Count: 1484
  - Unique: 301
  - Most frequent (top): Peugeot Partner Van Small (104 occurrences)

- **Product**
  - Count: 1484
  - Unique: 5
  - Most frequent (top): Diesel (1440 occurrences)

- **Odometer**
  - Count: 1484
  - Unique: 1359
  - Most frequent (top): 0 (114 occurrences)

- **Distance**
  - Count: 1484
  - Unique: 1187
  - Most frequent (top): 0 (250 occurrences)

- **Unit**
  - Count: 1482
  - Unique: 4
  - Most frequent (top): Miles (1265 occurrences)

- **MPG**
  - Count: 1482
  - Mean: 12.0377
  - Standard Deviation: 246.9147
  - Minimum: -9238.62
  - 25th Percentile: 3.1675
  - 50th Percentile (Median): 19.75
  - 75th Percentile: 29.885
  - Maximum: 824.85

### Notes

- The dataset has some missing values across several columns, which will be addressed during the data cleaning and preprocessing phase.
- The `MPG` column has some extreme values, which may need further investigation for outliers or data entry errors.

In [8]:
#1.5: Removing the unwanted columns but first let's look again the data we have
df.head()

Number Registration                              Details Product Odometer  \
0    NaN          NaN                                  NaN     NaN      NaN   
1   0001           U1  Jaguar XJL Portfolio V6 D Limousine  Diesel    67691   
2   0012      FL65FBA             Vauxhall CORSA  CDTI Car  Diesel     9329   
3   0013      MJ16XMY   Vauxhall Mokka 4x4 Utility Vehicle  Diesel    29064   
4   0016      BP16MKK                      Nissan Note Car  Diesel    42233   

  Distance   Unit    MPG  
0      NaN    NaN    NaN  
1        0  Miles   0.00  
2     2849  Miles  50.04  
3     5369  Miles  39.29  
4     1052  Miles  56.17

# Unwanted Columns:
1. Number:  because it has no affect on the fuel management it is just a index type
2. Registration: It could be useful for data management but not here

In [10]:
# Dropping the columns
df.drop(['Number', 'Registration'], axis = 1, inplace = True)
df.head()

Details Product Odometer Distance   Unit    MPG
0                                  NaN     NaN      NaN      NaN    NaN    NaN
1  Jaguar XJL Portfolio V6 D Limousine  Diesel    67691        0  Miles   0.00
2             Vauxhall CORSA  CDTI Car  Diesel     9329     2849  Miles  50.04
3   Vauxhall Mokka 4x4 Utility Vehicle  Diesel    29064     5369  Miles  39.29
4                      Nissan Note Car  Diesel    42233     1052  Miles  56.17

# 2. Data Cleaning

# 3. Feature Engineering